pkg/boot/parser.go

```diff
@@ -0,0 +1,267 @@
+		repNames := strings.Replace(trimmedLine, "'", "\"", -1)
+		names := strings.Split(repNames, "\"")
+		if len(names) > 1 {
+			name = names[1]
+		}
+	case "SET": // grub
+		if len(f) > 1 {
+			p.parseSearchHandleSet(f[1])
+		}
+	case "LABEL": // syslinux
+		p.state = syslinux
+		newEntry = true
+		name = trimmedLine[6:]
+		if name == "" {
+			name = "linux" // default for syslinux
+		}
+	case "DEFAULT": // syslinux
+		if len(f) > 1 {
+			label := strings.Join(f[1:], " ")
+			if !strings.HasSuffix(label, ".c32") {
+				p.defaultName = label
+			}
+		}
+	}
+
+	if newEntry {
+		p.entry = &Entry{
+			Name: name,
+			Type: Elf,
+		}
+	}
+}
+
+func (p *parser) parseSearchHandleSet(val string) {
+	val = strings.Replace(val, "\"", "", -1)
+	expr := strings.Split(val, "=")
+	if len(expr) != 2 {
+		return
+	}
+
+	switch expr[0] {
+	case "default":
+		index, err := strconv.Atoi(expr[1])
+		if err == nil {
+			p.defaultIndex = index
+		}
+	default:
+		// TODO: handle variables when grub conditionals are implemented
+		return
+	}
+}
+
+func (p *parser) parseGrubEntry(line string) {
+	trimmedLine := strings.TrimSpace(line)
+	f := strings.Fields(trimmedLine)
+	if len(f) == 0 {
+		return
+	}
+
+	switch f[0] {
+	case "}":
+		p.finishEntry()
+	case "multiboot":
+		p.entry.Type = Multiboot
+		p.entry.Modules = append(p.entry.Modules, NewModule(f[1], f[2:]))
+	case "module":
+		var filteredParams []string
+		for _, param := range f {
+			if param != "--nounzip" {
+				filteredParams = append(filteredParams, param)
+			}
+		}
+		p.entry.Modules = append(p.entry.Modules,
+			NewModule(filteredParams[1], filteredParams[2:]))
+	case "linux":
+		p.entry.Modules = append(p.entry.Modules, NewModule(f[1], f[2:]))
+	case "initrd":
+		p.entry.Modules = append(p.entry.Modules, NewModule(f[1], nil))
+	}
+}
+
+func (p *parser) parseSyslinuxEntry(line string) {
+	trimmedLine := strings.TrimSpace(line)
+	if len(trimmedLine) == 0 {
+		p.finishEntry()
+		return
+	}
+
+	f := strings.Fields(trimmedLine)
+	val := strings.Join(f[1:], " ")
+	switch strings.ToUpper(f[0]) {
+	case "LABEL": // new entry, finish this one and start a new one
+		p.finishEntry()
+		p.parseSearch(line)
+	case "MENU":
+		if len(f) > 1 {
+			switch strings.ToUpper(f[1]) {
+			case "LABEL":
+				tempName := strings.Join(f[2:], " ")
+				p.entry.Name = strings.Replace(tempName, "^", "", -1)
+			case "DEFAULT":
+				p.defaultName = p.entry.Name
+			}
+		}
+	case "LINUX", "KERNEL":
+		p.parseSyslinuxKernel(val)
+	case "INITRD":
+		p.entry.Modules = append(p.entry.Modules, NewModule(f[1], nil))
+	case "APPEND":
+		p.parseSyslinuxAppend(val)
+	}
+}
+
+func (p *parser) parseSyslinuxKernel(val string) {
+	if strings.HasSuffix(val, "mboot.c32") {
+		p.entry.Type = Multiboot
+	} else if strings.HasSuffix(val, ".c32") {
+		// skip this entry - not valid for kexec
+		p.state = search
+	} else {
+		p.entry.Modules = append(p.entry.Modules, NewModule(val, nil))
+	}
+}
+
+func (p *parser) parseSyslinuxAppend(val string) {
```

Comment:
*sigh* it looks like we're both working on a syslinux parser. #489 is working towards a pkg/pxe (and a nicer parser, once I tear apart the internals and add more tests).

I'm a bit confused as to why you have two different parsers in one structure? Grub and syslinux?

Reply:
two parsers in one struct:because the parts we care about are so similar that it does not matter: a limited incorrect thing covers almost all cases. And as you point out the ramifications of the include directive are huge since it can cover including fairly arbitrary things. And, the grub command language, which is now turing complete, is supposed to become much simpler in 2018 -- we had a discussion with phicoder about this. We proposed JSON or YAML and that seemed acceptable. 

So, for now, dead simple and not very capable parser is a reasonable solution; the fully capable parser for grub is a really big thing and will take a long time to get right. Something that's clearly limited is almost always better than something that claims to be fully capable but is not. 

A huge amount of syslinux and grub config involves menus and colors and selections and a command line -- all because we're the first thing running from the BIOS. How much of that do we really need given we are invoking a boot command from a shell under a Linux kernel? Almost none of it. We just need to know the names of the boot choices and the kernel/initramfs/cmdline that go with them. So, sure, write a full parser but be mindful of how much of it we need vs. how much of it we actually implement.

Reply:
I didnt say anything about a full parser... Neither of us have been working on a full one. Mine covers one more directive than this one.

We talked on slack and decided to eventually converge these two parsers, but keep working on it for now.

Reply:
got it. I got confused.